cmake/onnxruntime_python.cmake

```diff
@@ -100,7 +80,27 @@ target_link_libraries(onnxruntime_pybind11_state PRIVATE
     onnxruntime_providers
     onnxruntime_util
     ${onnxruntime_tvm_libs}
-    onnxruntime_framework
+)
+
+if (onnxruntime_ENABLE_TRAINING)
+  if (onnxruntime_ENABLE_TRAINING_TORCH_INTEROP)
+    list(APPEND onnxruntime_pybind11_state_link_targets onnxruntime_interop_torch)
```

Comment:
What I care most is if we have a clear picture of which onnxruntime component depends on which. Like does your onnxruntime_interop_torch depend on onnxruntime_framework or the reverse? And we must carefully avoid making any circular dependency loop.

That's why here again and again I emphasize the order of the libraries is wrong. You put them in wrong order. 

So, please think about it: 
1.  Does onnxruntime_interop_torch use any function from onnxruntime_framework?
2.  The reverse: does onnxruntime_framework use any function from onnxruntime_interop_torch ?

Reply:
The answer is, onnxruntime_interop_torch depend on onnxruntime_framework. onnxruntime_framework should not use any fuynction from onnxruntime_interop_torch .

Reply:
Any further concerns?

Reply:
Then how about onnxruntime_python_interface?

Reply:
onnxruntime_interop_torch  depends on onnxruntime_framework depends on onnxruntime_python_interface

Reply:
As we talked offline, will think about how to make it clear between onnxruntime_framework  and onnxruntime_python_interface

Reply:
Remove both onnxruntime_interop_torch  and onnxruntime_python_interface.

Duplicate dlpack_python.h/cc implementation, then the python dependency become simpler. 

> One for pybindstate usage. this will only include them when training is enabled (because training will use dlpack python interface). For other lib for example libonnxruntime.so, it won't include this API, so no python.h dependency introduced. 

So overall, if TRAINING is enabled and torch interop is disabled, only pybindstate use python.h.

> One  for framework usage when torch interop is enabled. 

So overall, if TRAINING is enabled and torch interop is eabled, pybindstate use python.h, and framework use python.h too. 
libonnxruntime still will fail, but  added switched to disable the torch interop once we found --enable_shared_lib is given in build command. This is the safest way so far.